### PR TITLE
Socket connection in python fuzzing

### LINKED ISSUE #1856: Add socket connection between UnitTestBot and PythonExecutor
**Description**

Now we can run approximate 200 fuzzed values but it is too little. Main reason is a long initialization of python interpreter so possible solution is creation of one python executor which will be connected with main kotlin program.

**Expected behavior**

Test generation is faster then now

### PR DESCRIPTION
**Pull request by tamarinvs19**
## Description

Added socket connection between python and kotlin. Now it works 10 times faster then previous verision.

Fixes #1856

### Manual tests

Generate tests for python function, for example, with many branches.

## Self-check list

Check off the item if the statement is true. Hint: [x] is a marked item.

Please do not delete the list or its items.

- [x] I've set the proper **labels** for my PR (at least, for category and component).
- [x] PR **title** and **description** are clear and intelligible.
- [x] I've added enough **comments** to my code, particularly in hard-to-understand areas.
- [ ] The functionality I've repaired, changed or added is covered with **automated tests**.
- [x] **Manual tests** have been provided optionally.
- [ ] The **documentation** for the functionality I've been working on is up-to-date.

### CODE REVIEW

**Review by Markoutte — COMMENTED** [review]
At the moment an exception is thrown, or no tests generated for this example:

```python
def test3(x: str) -> bool:
    if x[0] == 'b':
        if x[1] == 'a':
            if x[2] == 'd':
                if x[3] == '!':
                    raise RuntimeError("Very bad")
    return True
```

File: utbot-python/src/main/kotlin/org/utbot/python/PythonEngine.kt [hunk 1]
```diff
@@ -146,108 +150,147 @@ class PythonEngine(
...
-                        "Error evaluation: ${evaluationResult.status}, ${evaluationResult.message}",
-                        Throwable(evaluationResult.stackTrace.joinToString("\n"))
-                    )
-                    logger.debug(evaluationResult.stackTrace.joinToString("\n"))
-                    emit(InvalidExecution(utError))
-                    PythonFeedback(control = Control.PASS)
+        ServerSocket(0).use { serverSocket ->
+            logger.info { "Server port: ${serverSocket.localPort}" }
+            val processStartTime = System.currentTimeMillis()
+//                val outputFile = "/home/vyacheslav/utbot_profile"
```
[the remark this answers is not in the document]
Comment (tamarinvs19) [reply utbot-python/src/main/kotlin/org/utbot/python/PythonEngine.kt]: Fixed

**Review by Markoutte — APPROVED** [review]

File: utbot-python/src/main/kotlin/org/utbot/python/fuzzing/PythonApi.kt [hunk 2]
```diff
@@ -40,9 +42,16 @@ class PythonMethodDescription(
...
+sealed interface FuzzingExecutionFeedback
+class ValidExecution(val utFuzzedExecution: UtFuzzedExecution): FuzzingExecutionFeedback
+class InvalidExecution(val utError: UtError): FuzzingExecutionFeedback
+class TypeErrorFeedback(val message: String) : FuzzingExecutionFeedback
+class ArgumentsTypeErrorFeedback(val message: String) : FuzzingExecutionFeedback
+
 data class PythonFeedback(
     override val control: Control = Control.CONTINUE,
     val result: Trie.Node<Instruction> = Trie.emptyNode(),
+    val executionFeedback: FuzzingExecutionFeedback?
```
Comment (Markoutte) [inline utbot-python/src/main/kotlin/org/utbot/python/fuzzing/PythonApi.kt:54]: This feedback is used as a part of equals/hashCode, therefore 2 same paths with different result will be recognized as different feedbacks. Is it intented?

File: utbot-python/src/main/kotlin/org/utbot/python/newtyping/utils/Utils.kt [hunk 3]
```diff
@@ -1,5 +1,16 @@
 package org.utbot.python.newtyping.utils
 
+import kotlin.random.Random
+
 fun getOffsetLine(sourceFileContent: String, offset: Int): Int {
     return sourceFileContent.take(offset).count { it == '\n' } + 1
 }
+
+fun <T> weightedRandom(elems: List<T>, weights: List<Double>): T {
```
Comment (Markoutte) [inline utbot-python/src/main/kotlin/org/utbot/python/newtyping/utils/Utils.kt:9]: You can find similar util method in RandomExtension.kt (in module utbot-fuzzing) with name Random.chooseOne. Also, please not, that random in your case doesn't use seed, therefore, every run will be different, that usually is undesirable.

**Review by Markoutte — CHANGES_REQUESTED** [review]

File: utbot-python/src/main/kotlin/org/utbot/python/newtyping/utils/Utils.kt [hunk 4]
```diff
@@ -8,9 +9,6 @@ fun getOffsetLine(sourceFileContent: String, offset: Int): Int {
 
 fun <T> weightedRandom(elems: List<T>, weights: List<Double>): T {
     val sum = weights.sum()
-    val borders = weights.fold(emptyList<Double>() to 0.0) { (list, partialSum), cur ->
-        (list + (partialSum + cur) / sum) to partialSum + cur
-    }.first
-    val value = Random.nextDouble()
-    return elems[borders.indexOfFirst { it >= value }]
+    val index = Random.chooseOne(weights.map { it / sum }.toDoubleArray())
```
Comment (Markoutte) [inline utbot-python/src/main/kotlin/org/utbot/python/newtyping/utils/Utils.kt:12]: The random is still so random. Also, you can pass to the chooseOne just `weights.toDoubleArray`. The method does `sum` itself, so, there's no need to do before.

File: utbot-python/src/main/kotlin/org/utbot/python/PythonEngine.kt [hunk 5]
```diff
@@ -52,6 +53,18 @@ class PythonEngine(
     private val pythonTypeStorage: PythonTypeStorage,
 ) {
 
+    private val cache = mutableMapOf<Pair<PythonMethodDescription, List<PythonTreeWrapper>>, PythonExecutionResult>()
+
+    private fun addExecutionToCache(key: Pair<PythonMethodDescription, List<PythonTreeWrapper>>, result: PythonExecutionResult) {
```
Comment (Markoutte) [inline utbot-python/src/main/kotlin/org/utbot/python/PythonEngine.kt:58]: I'd recommend extracting the cache into a separate class to facilitate further work with it.

**Review by Markoutte — CHANGES_REQUESTED** [review]

**Review by Markoutte — APPROVED** [review]

**Pull request merged**